src/librustc/ty/layout.rs

```diff
@@ -1125,354 +835,760 @@ impl<'tcx> fmt::Display for LayoutError<'tcx> {
-                    StructKind::AlwaysSizedUnivariant, ty)?;
-                Univariant { variant: st, non_zero: false }
+                    StructKind::AlwaysSized)?
             }
 
             ty::TyTuple(tys, _) => {
                 let kind = if tys.len() == 0 {
-                    StructKind::AlwaysSizedUnivariant
+                    StructKind::AlwaysSized
                 } else {
-                    StructKind::MaybeUnsizedUnivariant
+                    StructKind::MaybeUnsized
                 };
 
-                let st = Struct::new(dl,
-                    &tys.iter().map(|ty| ty.layout(tcx, param_env))
-                      .collect::<Result<Vec<_>, _>>()?,
-                    &ReprOptions::default(), kind, ty)?;
-                Univariant { variant: st, non_zero: false }
+                univariant(&tys.iter().map(|ty| cx.layout_of(ty)).collect::<Result<Vec<_>, _>>()?,
+                    &ReprOptions::default(), kind)?
             }
 
             // SIMD vector types.
             ty::TyAdt(def, ..) if def.repr.simd() => {
-                let element = ty.simd_type(tcx);
-                match *element.layout(tcx, param_env)? {
-                    Scalar { value, .. } => {
-                        return success(Vector {
-                            element: value,
-                            count: ty.simd_size(tcx) as u64
-                        });
-                    }
+                let count = ty.simd_size(tcx) as u64;
+                let element = cx.layout_of(ty.simd_type(tcx))?;
+                match element.abi {
+                    Abi::Scalar(_) => {}
                     _ => {
                         tcx.sess.fatal(&format!("monomorphising SIMD type `{}` with \
                                                 a non-machine element type `{}`",
-                                                ty, element));
+                                                ty, element.ty));
                     }
                 }
+                let size = element.size.checked_mul(count, dl)
+                    .ok_or(LayoutError::SizeOverflow(ty))?;
+                let align = dl.vector_align(size);
+                let size = size.abi_align(align);
+
+                tcx.intern_layout(CachedLayout {
+                    variants: Variants::Single { index: 0 },
+                    fields: FieldPlacement::Array {
+                        stride: element.size,
+                        count
+                    },
+                    abi: Abi::Vector,
+                    size,
+                    align,
+                    primitive_align: align
+                })
             }
 
             // ADTs.
             ty::TyAdt(def, substs) => {
-                if def.variants.is_empty() {
-                    // Uninhabitable; represent as unit
-                    // (Typechecking will reject discriminant-sizing attrs.)
+                // Cache the field layouts.
```

Comment:
codestyle bomb: it might be a better idea to split this into a separate closure within `compute_uncached` for each layout mode (union, struct, niche-filling, discriminant) instead of having all of this logic mixed up within 5 nesting levels.